src/stages/stage3_events.c

```diff
@@ -741,939 +81,689 @@ Boss* stage3_spawn_scuttle(cmplx pos) {
+		if(!p2) {
+			p2 = 5 + (0.93 + 0.01 * x) * I;
+		}
+		INVOKE_TASK(side_swirl, ARGS.start_pos, ARGS.p0, ARGS.p1, p2);
+		WAIT(interval);
 	}
+}
 
-	// night ignite balls
-	if(creal(e->args[3]) && global.diff > D_Easy) {
-		FROM_TO(300, 1000000, 180) {
-			int cnt = 5, i;
-			for(i = 0; i < cnt; ++i) {
-				PROJECTILE(
-					.proto = pp_ball,
-					.pos = e->pos,
-					.color = RGBA(0.5, 1.0, 0.5, 0),
-					.rule = accelerated,
-					.args = {
-						0, 0.02 * cexp(I*i*2*M_PI/cnt)
-					},
-				);
+// fairies
 
-				if(global.diff > D_Hard) {
-					PROJECTILE(
-						.proto = pp_ball,
-						.pos = e->pos,
-						.color = RGBA(1.0, 1.0, 0.5, 0),
-						.rule = accelerated,
-						.args = {
-							0, 0.01 * cexp(I*i*2*M_PI/cnt)
-						},
-					);
-				}
-			}
+TASK(little_fairy, { cmplx pos; cmplx target_pos; int danmaku_type; int side; }) {
+	// contains stage3_slavefairy and stage3_slavefairy2
+	Enemy *e = TASK_BIND_UNBOXED(create_enemy1c(ARGS.pos, 900, Fairy, NULL, 0));
 
-			// FIXME: better sound
-			play_sound("shot_special1");
-		}
-	}
+	// fade-in
+	e->alpha = 0;
 
-	return 1;
-}
+	INVOKE_TASK_WHEN(&e->events.killed, common_drop_items, &e->pos, {
+			.points = 3,
+			.power = 2,
+			});
 
-void wriggle_moonlight_rocket(Boss *boss, int time) {
-	int i, j, cnt = 1 + global.diff;
-	TIMER(&time)
+	e->move.attraction_point = ARGS.target_pos;
 
-	AT(EVENT_DEATH) {
-		enemy_kill_all(&global.enemies);
-		return;
-	}
+	int shot_interval = 1;
+	int intensity = difficulty_value(90, 70, 50, 30);
 
-	if(time < 0)
-		GO_TO(boss, VIEWPORT_W/2 + VIEWPORT_H*I/2.5, 0.05)
-	else if(time == 0) {
-		for(j = -1; j < 2; j += 2) for(i = 0; i < cnt; ++i)
-			create_enemy3c(boss->pos, ENEMY_IMMUNE, wriggle_slave_visual, wriggle_spell_slave, add_ref(boss), i*2*M_PI/cnt, j);
-	}
-}
+	switch(ARGS.danmaku_type) {
+		case 1:
+			e->move.attraction = 0.05;
+			WAIT(60);
 
-static int wriggle_ignite_laserbullet(Projectile *p, int time) {
-	if(time == EVENT_DEATH) {
-		free_ref(p->args[0]);
-		return ACTION_ACK;
-	} else if(time < 0) {
-		return ACTION_ACK;
-	}
+			for (int i = 0; i < intensity; ++i) {
+				float a = global.timer * 0.5;
+				cmplx dir = cdir(a);
 
-	Laser *laser = (Laser*)REF(p->args[0]);
+				// fire out danmaku in all directions in a spiral-ish pattern
+				PROJECTILE(
+						.proto = pp_wave,
+						.pos = e->pos + dir * 10,
+						.color = (i % 2) ? RGB(1.0, 0.3, 0.3) : RGB(0.3, 0.3, 1.0),
+						.move = move_accelerated(dir, dir * 0.025),
+						);
+
+				// danmaku that only shows up above Easy difficulty
+				if(global.diff > D_Easy) {
+					PROJECTILE(
+							.proto = pp_ball,
+							.pos = e->pos + dir * 10,
+							.color = RGB(1.0, 0.6, 0.3),
+							.move = move_linear(dir * (1.0 + 0.5 * sin(a))),
+							);
+				}
 
-	if(laser) {
-		p->args[3] = laser->prule(laser, time - p->args[1]) - p->pos;
-	}
+				play_sound("shot1");
+				WAIT(shot_interval);
+			}
 
-	p->angle = carg(p->args[3]);
-	p->pos = p->pos + p->args[3];
+		case 2:
```

Comment:
I think there's a subtle bug here. `case 1` doesn't `break`, so it will fall through to `case 2`. Another reason to avoid the switch.